File: DSA/fundamentals/arrays_and_lists/part0/cpp/Judge.cpp
#include <iostream>
#include <vector>
#include "Solution.cpp"
using namespace std;

class Judge {
public:
    void runTests() {
        vector<vector<int>> testCasesA = {
            {1, 2, 3, 4, 5},
            {1, 2, 3, 4, 5}
        };
        vector<int> testCasesK = {1, 0};
        
        runTests(testCasesA, testCasesK);
    }

    void runTests(vector<vector<int>>& testCasesA, vector<int>& testCasesK) {
        int passes = 0;
        for (int i = 0; i < testCasesA.size(); i++) {
            vector<int> expected = testCaseSolutionOutput(testCasesA[i], testCasesK[i]);
            if (test(testCasesA[i], testCasesK[i], expected)) {
                passes++;
            }
        }
        cout << "Passed " << passes << " of " << testCasesA.size() << " tests." << endl;
    }

private:
    bool test(vector<int>& a, int k, vector<int>& expected) {
        vector<int> originalA = a;
        Solution sol;
        sol.incrementArray(a, k);
        bool passed = (a == expected);
        
        cout << "a: [";
        for (int i = 0; i < originalA.size(); i++) {
            cout << originalA[i];
            if (i < originalA.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
        
        cout << "k: " << k << endl;
        
        cout << "Expected Output: [";
        for (int i = 0; i < expected.size(); i++) {
            cout << expected[i];
            if (i < expected.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
        
        cout << "Your Output: [";
        for (int i = 0; i < a.size(); i++) {
            cout << a[i];
            if (i < a.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
        
        cout << (passed ? "Test PASSED" : "Test FAILED") << endl;
        cout << "-----------" << endl;
        return passed;
    }

    vector<int> testCaseSolutionOutput(vector<int> a, int k) {
        for (int i = 0; i < a.size(); i++) {
            a[i] += k;
        }
        return a;
    }
};

File: DSA/fundamentals/arrays_and_lists/part0/java/Judge.java
package DSA.fundamentals.arrays_and_lists.part0.java;

import java.util.*;

public class Judge {

    public void runTests() {
        runTests(
                new int[][]{
                    {1, 2, 3, 4, 5},
                    {1, 2, 3, 4, 5},},
                new int[]{
                    1,
                    0,}
        );
    }

    public void runTests(int[][] testCasesA, int[] testCasesK) {
        int passes = 0;
        for (int i = 0; i < testCasesA.length; i++) {
            int[] copyUser = Arrays.copyOf(testCasesA[i], testCasesA[i].length);
            int[] copyExpected = Arrays.copyOf(testCasesA[i], testCasesA[i].length);
            if (test(copyUser, testCasesK[i], testCaseSolutionOutput(copyExpected, testCasesK[i]))) {
                passes++;
            }
        }
        System.out.println("Passed " + passes + " of " + testCasesA.length + " tests.");
    }

    private boolean test(int[] a, int k, int[] expected) {
        int[] originalA = Arrays.copyOf(a, a.length);
        Solution sol = new Solution();
        sol.incrementArray(a, k);
        boolean passed = Arrays.equals(a, expected);

        System.out.println("a: " + Arrays.toString(originalA));
        System.out.println("k: " + k);
        System.out.println("Expected Output: " + Arrays.toString(expected));
        System.out.println("Your Output: " + Arrays.toString(a));
        System.out.println(passed ? "Test PASSED" : "Test FAILED");
        System.out.println("-----------");
        return passed;
    }

    private int[] testCaseSolutionOutput(int[] a, int k) {
        for (int i = 0; i < a.length; i++) {
            a[i] += k;
        }
        return a;
    }

}
